### PR TITLE
Support for writes in FullAcid and InsertOnly tables.

### PR DESCRIPTION
**Pull request by citrusraj**

### CODE REVIEW

File: src/main/scala/com/qubole/spark/hiveacid/HiveAcidMetadata.scala [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    if (sparkSession.sessionState.conf.caseSensitiveAnalysis) {
+      field.name
+    } else {
+      field.name.toLowerCase(Locale.ROOT)
+    }
+  }
+}
+
+object HiveAcidMetadata {
+  val DEFAULT_DATABASE = "default"
```
Comment (citrusraj) [inline src/main/scala/com/qubole/spark/hiveacid/HiveAcidMetadata.scala]: Support for default database in absence of mention about it.

File: README.md [hunk 2]
```diff
@@ -129,37 +200,26 @@ Read more about [sbt release](https://github.com/sbt/sbt-release)
...
+3. The files in the snapshot needs to be protected till the RDD is in use. By design concurrent reads and writes on the Hive ACID works with the help of locks, where every client (across multiple engines) that is operating on ACID tables is expected to acquire locks for the duration of reads and writes. The lifetime of RDD can be very long, to avoid blocking other operations like inserts this datasource _DOES NOT_ acquire lock but uses an alternative mechanism to protect reads. Other way the snapshot can be protected is by making sure the files in the snapshot are not deleted while in use. For the current datasoure any table on which Spark is operating `Automatic Cleanup` should be disabled. This makes sure that `compaction` is performed but cleaner does not clean any file. To disable automatic cleaner on table
 
-Once you have disabled automatic compaction either globally or on a particular set of tables, you can chose to run compaction manually at a desired time when you know there are no readers reading these acid tables, using an ALTER TABLE command:
+         ALTER TABLE <> SET TBLPROPERTIES ("NO_CLEANER"="true")
```
Comment (amoghmargoor) [inline README.md]: Is this Qubole Specific ?

File: README.md [hunk 3]
```diff
@@ -129,37 +200,26 @@ Read more about [sbt release](https://github.com/sbt/sbt-release)
...
 
-        ALTER TABLE table_name [PARTITION (partition_key = 'partition_value' [, ...])] COMPACT 'compaction_type'[AND WAIT] [WITH OVERWRITE TBLPROPERTIES ("property"="value" [, ...])];
+	When the table is not in use cleaner can be enabled and all the files that needs cleaned will get queued up for cleaner. Disabling cleaner has no performance implication on reads/writes, only the storage requirement will be higher as system may hold on to all the snapshots in use.
 
-compaction_type are either `MAJOR` or `MINOR`
 
-More details on the above commands and their variations available [here](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL).
+4. Note that even though reads are protected admin operation like `TRUNCATE` `ALTER TABLE DROP COLUMN` and `DROP` have no protected as they clean files with intevention from cleaner. These operations should be performed when Spark is not using the table.
```
Comment (amoghmargoor) [inline README.md]: change from `have no protected` to  `have no protection`

**Review by amoghmargoor — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (citrusraj) [reply README.md]: fixed

File: README.md (hunk 3 above)
Comment (citrusraj) [reply README.md]: fixed

File: src/main/scala/com/qubole/spark/hiveacid/rdd/HiveAcidRDD.scala [hunk 4]
```diff
@@ -238,62 +237,39 @@ class Hive3RDD[K, V](
...
+      HiveAcidRDD.addLocalConfiguration(
         new SimpleDateFormat("yyyyMMddHHmmss", Locale.US).format(createTime),
         context.stageId, theSplit.index, context.attemptNumber, jobConf)
 
       reader =
         try {
-          inputFormat.getRecordReader(split.inputSplit.value, jobConf, Reporter.NULL)
+          // Underlying code is not MT safe. Synchronize
+          // while creating record reader
+          HiveAcidRDD.RECORD_READER_INIT_LOCK.synchronized {
```
Comment (citrusraj) [inline src/main/scala/com/qubole/spark/hiveacid/rdd/HiveAcidRDD.scala:270]: Fix for issue https://github.com/qubole/spark-acid/issues/19

**Pull request merged**